pallets/msa/src/lib.rs

```diff
@@ -957,6 +972,74 @@ impl<T: Config> Pallet<T> {
 		}
 		Ok(Some(schemas.into_inner()))
 	}
+
+	/// Adds a signature to the PayloadSignatureRegistry based on a virtual "bucket" grouping.
+	/// Check that mortality_block is within bounds. If so, proceed and add the new entry.
+	/// Raises `SignatureAlreadySubmitted` if the bucket-signature double key exists in the
+	/// registry.
+	pub fn register_signature(
+		signature: &MultiSignature,
+		mortality_block: T::BlockNumber,
+	) -> DispatchResult {
+		let current_block = frame_system::Pallet::<T>::block_number();
+
+		let max_lifetime = Self::mortality_block_limit(current_block);
+		if max_lifetime < mortality_block {
+			Err(Error::<T>::ProofNotYetValid.into())
+		} else if current_block >= mortality_block {
+			Err(Error::<T>::ProofHasExpired.into())
+		} else {
+			let bucket_num = Self::bucket_for(mortality_block.into());
+			<PayloadSignatureRegistry<T>>::try_mutate(
+				bucket_num,
+				signature,
+				|maybe_mortality_block| -> DispatchResult {
+					ensure!(maybe_mortality_block.is_none(), Error::<T>::SignatureAlreadySubmitted);
+					*maybe_mortality_block = Some(mortality_block);
+					Ok(())
+				},
+			)
+		}
+	}
+
+	// Check if enough blocks have passed to reset bucket mortality storage.
+	// If so:
+	//     1. delete all the stored bucket/signature alues with key1 = bucket num
+	//	   2. add the WeightInfo proportional to the storage read/writes to the block weight
+	// If not, don't do anything.
+	fn reset_virtual_bucket_if_needed(current_block: T::BlockNumber) -> Weight {
+		let bucket_size: T::BlockNumber = T::MortalityBucketSize::get().into();
+
+		// Now that we're in a "new bucket block set"
+		if current_block % bucket_size != T::BlockNumber::zero() {
+			return T::WeightInfo::on_initialize(0 as u32)
+		}
+
+		// Clear the previous bucket block set
+		let bucket_num = Self::bucket_for(current_block - T::BlockNumber::one());
+		let multi_removal_result = <PayloadSignatureRegistry<T>>::clear_prefix(
+			bucket_num,
+			T::MaxSignaturesPerBucket::get(),
+			None,
+		);
+		T::WeightInfo::on_initialize(multi_removal_result.unique)
+	}
+
+	// The furthest in the future a mortality_block value is allowed
+	// to be for current_block
+	// This is calculated to be past the risk of a replay attack
+	fn mortality_block_limit(current_block: T::BlockNumber) -> T::BlockNumber {
+		let bucket_size: T::BlockNumber = T::MortalityBucketSize::get().into();
+		let num_buckets: T::BlockNumber = T::NumberOfBuckets::get().into();
+		current_block + (bucket_size * (num_buckets - T::BlockNumber::one()))
```

Comment:
Just want to mentioned that this seems to be based on assumption that always `num_buckets > 1`

Reply:
it is always > 1